python/paddle/distributed/auto_parallel/static/tuner/pir_rule_based_tuner.py

```diff
@@ -0,0 +1,1487 @@
+            # print(f"compare operand, not op")
+            # compare tensor, from tensor to op
+
+            # as input for op node
+            src_as_input_ops = src.all_used_ops()
+            tgt_as_input_ops = tgt.all_used_ops()
+            # if len(src_as_input_ops) != len(tgt_as_input_ops):
+            #     print(f"src_as_input_ops not match")
+            #     not_matched = True
+            #     return
+            # todo: process src_as_input_ops < tgt_as_input_ops
+            if len(src_as_input_ops) > len(tgt_as_input_ops):
+                not_matched = True
+                return
+            if len(src_as_input_ops) == len(tgt_as_input_ops):
+                for idx, src_as_input_op in enumerate(src_as_input_ops):
+                    src_as_input_op_id = (
+                        src_as_input_op.get_parent_block().ops.index(
+                            src_as_input_op
+                        )
+                    )
+                    if src_as_input_op_id in result.keys():
+                        continue
+
+                    tgt_as_input_op = tgt_as_input_ops[idx]
+                    _match_core(src_as_input_op, tgt_as_input_op, is_op=True)
+
+            # as output for op node
+            src_as_output_op = src.get_defining_op()
+            tgt_as_output_op = tgt.get_defining_op()
+            if src_as_output_op is not None and tgt_as_output_op is not None:
+                src_as_output_op_id = (
+                    src_as_output_op.get_parent_block().ops.index(
+                        src_as_output_op
+                    )
+                )
+                if src_as_output_op_id not in result.keys():
+                    _match_core(src_as_output_op, tgt_as_output_op, is_op=True)
+
+    results = []
+    result = {}
+    matched_ids = set()
+    matched_op_node_ids = set()
+
+    # starts with a op node
+    src_ops = pattern.pir_program.global_block().ops
+    for op in src_ops:
+        if op.name() != "pd_op.data" and op.name() != "builtin.parameter":
+            src_start_op = op
+            break
+    # src_start_op = src_ops[0] # to be done, need to check pattern start op
+    assert src_start_op is not None, "src_start_op is none"
+    # print(f"start op of this pattern is: {src_start_op}")
+
+    tgt_ops = pir_program.global_block().ops
+    for idx, tgt_op in enumerate(tgt_ops):
+        if tgt_op.name() == src_start_op.name():
+            # print(f"program op index is {idx}")
+            # print(f"start match core")
+            not_matched = False
+            result = {}
+            _match_core(src_start_op, tgt_op, is_op=True)
+            if not not_matched:
+                # print(f"matched one, check whether to append")
+                need_to_append = True
+                for value in result.values():
+                    if value in matched_op_node_ids:
+                        result = {}
+                        need_to_append = False
+                        break
```

Comment:
Should this `break` be removed, such as `pattern=(matmul, add, matmul, add,)` and `prog=(matmul, add, matmul, add, matmul, add, Relu, ...)` will it cause a matching interruption?